middleware/request/retry.go

```diff
@@ -10,19 +10,22 @@ import (
 // NewRetryMiddleware returns a new retry middleware that wraps the provided `next` request handler.
 // The middleware retries the request a maximum of `maxRetries` times with a delay of `interval` between each retry.
 // If the request succeeds at any retry, the middleware returns `nil`. If all retries fail, it returns the last error encountered.
-func NewRetryMiddleware(maxRetries int, interval time.Duration) func(next wasabi.RequestHandler) wasabi.RequestHandler {
+func NewRetryMiddleware(retryConfig RetryConfig) func(next wasabi.RequestHandler) wasabi.RequestHandler {
```

Comment:
@shan-96 what do you think if we'll accept here function as argument? that will take one parameter current retry counter and return duration? we can keep config inside closure...

Reply:
something like this 

```golang
func ExponentialRetryConfig(seed time.Duration, delayFactor int) func (int) time.Duuration{
 	return func (iteration int) time.Duration {
        return seed * (time.Duration(math.Pow(float64(delayFactor), float64(iteration))))v
    }
 }
```

Reply:
The benefit I see that, it'll provide possibility to add any policy if for some reason start one is not enough

Reply:
you know my initial commit did something similar. Yes can do this

Reply:
@ksysoev had to still keep retry config because maxRetries needs to be available for outer methods. Let me know you have idea to work around thatt

Reply:
@shan-96 but in this case we can just pass as arguments without additional structure, like:

```
request.NewRetryMiddleware(3, request.ExponentialRetryPolicy())
```

will it be simpler?

Reply:
yes and I have just wrapped maxretries and the higher order function in the config in the new commit
```go
type RetryConfig struct {
	getRetryInterval GetRetryInterval
	maxRetries       int
}
```

Reply:
@shan-96 this structure is impossible to use outside of `request` package, all fields are private. Do we really need to introduce additional structure? just parameters looks simpler to me

Reply:
also as an option we can ditch maxRetries as well... if duration is that returned by the policy is `0`, we can stop retrying.. what doe you think?

Reply:
getting rid of max reties also seems like a good idea. We can keep retrying until the middleware's `shouldRetry` function returns false or something like that

Reply:
I meant something like this:

```golang
for {
    ...
    intr = retryConfig.getRetryInterval(i)
    if intr == 0 {
       break
    }
    timer.Reset(intr)
    ...
}
```

or if you don't like special value, we can return tuple like 


```golang
for {
    ...
    intr, ok = retryConfig.getRetryInterval(i)
    if !ok {
       break
    }
    timer.Reset(intr)
    ...
}
```

Reply:
What if I keep the logic of getting the interval and getting the breaking condition separate? Let me add a commit